### PR TITLE
feat: add `domain_ou` support for windows customization

### LINKED ISSUE #2180: Add support for domain_OU windows customization
### Community Guidelines

- [X] I have read and agree to the [HashiCorp Community Guidelines ](https://www.hashicorp.com/community-guidelines).
- [X] Vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue initial description to help the maintainers prioritize.
- [X] Do not leave "+1" or other comments that do not add relevant information or questions.
- [X] If you are interested in working on this issue or have submitted a pull request, please leave a comment.

### Description

there is realted https://github.com/hashicorp/terraform-provider-vsphere/issues/996 . however after that domain_OU parameter has been released in vcenter 8.

### Use Case(s)

add VM into specific during creation via single param i.e.
```hcl
resource "vsphere_virtual_machine" "vm" {
  # ... other configuration ...
  clone {
    # ... other configuration ...
    customize {
      # ... other configuration ...
      windows_options {
        domain_ou             = "OU=Sandbox,OU=Workstations,DC=example,DC=com"
      }
    }
  }
}
```
instead kludges like

```hcl
  provisioner "remote-exec" {
   inline = [
     "powershell.exe Add-Computer -DomainName 'example.com' -OUPath 'OU=Sandbox,OU=Workstations,DC=example,DC=com' -Credential (New-Object System.Management.Automation.PSCredential('example\\DomainJoin', (ConvertTo-SecureString 'mysecretpassword' -AsPlainText -Force))) -Restart -Force"
   ]
    connection {
     type     = "winrm"
     user     = "Administrator"
     password = "anothersecretpassword"
     use_ntlm = true
     host     = self.default_ip_address
    }
  }

  provisioner "remote-exec" {
   when = destroy
   inline = [
     "powershell.exe Remove-Computer -UnjoinDomaincredential (New-Object System.Management.Automation.PSCredential('example\\DomainJoin', (ConvertTo-SecureString 'mysecretpassword' -AsPlainText -Force))) -Force"
   ]
   connection {
     type     = "winrm"
     user     = "DomainJoin"
     password = "mysecretpassword"
     use_ntlm = true
     host     = self.default_ip_address 
    }
  }
} 
```

### Potential Terraform Provider Configuration

```hcl
resource "vsphere_virtual_machine" "vm" {
  # ... other configuration ...
  clone {
    # ... other configuration ...
    customize {
      # ... other configuration ...
      windows_options {
        domain_ou             = "OU=Sandbox,OU=Workstations,DC=example,DC=com"
      }
    }
  }
}
```

### References

* https://github.com/hashicorp/terraform-provider-vsphere/issues/996
* https://developer.vmware.com/apis/vsphere-automation/latest/vcenter/data-structures/Guest/Domain/
* https://learn.microsoft.com/en-us/windows-hardware/customize/desktop/unattend/microsoft-windows-unattendedjoin-identification-machineobjectou

### PR DESCRIPTION
**Pull request by ultral**
### Description

allow to put VMs into specific OU. closes #2180

### Acceptance tests
- [ ] Have you added an acceptance test for the functionality being added?
- [ ] Have you run the acceptance tests on this branch?

### Release Note

```release-note
Add support for domain_OU windows customization
```
### References

* https://github.com/hashicorp/terraform-provider-vsphere/issues/2180
* https://github.com/hashicorp/terraform-provider-vsphere/issues/996
* https://github.com/hashicorp/terraform-provider-vsphere/issues/2130

### CODE REVIEW

**Comment by tenthirtyam** [comment]
Marked as drafted until the version conditions are included.

File: vsphere/internal/helper/guestoscustomizations/customizations_helper.go [hunk 1]
```diff
@@ -429,6 +436,7 @@ func flattenWindowsOptions(customizationPrep *types.CustomizationSysprep) ([]map
 		winOptionsData["domain_admin_password"] = customizationPrep.Identification.DomainAdminPassword.Value
 	}
 	winOptionsData["join_domain"] = customizationPrep.Identification.JoinDomain
+	winOptionsData["domain_OU"] = customizationPrep.Identification.DomainOU
```
[the remark this answers is not in the document]
Comment (ultral) [reply vsphere/internal/helper/guestoscustomizations/customizations_helper.go:439]: @tenthirtyam as i can see the only one option to detect version is to use `*govmomi.Client`.  it means that the client object must be somehow passed to the `SpecSchema` function. I'm not really good in that plugin code, but for me it looks like leaking abstraction. do you have any other ideas how to detect version in other way in that place?

**Comment by ultral** [comment]
@tenthirtyam I've added version check. however, I'm not 100% that it's ideal solution. Could you please take a look? are there any other ways to check the version?

File: vsphere/internal/helper/guestoscustomizations/customizations_helper.go [hunk 2]
```diff
@@ -323,7 +331,8 @@ func FlattenGuestOsCustomizationSpec(d *schema.ResourceData, specItem *types.Cus
 			specData["windows_sysprep_text"] = sysprepText
 		} else {
 			specItemWinOptions := specItem.Spec.Identity.(*types.CustomizationSysprep)
-			windowsOptions, err := flattenWindowsOptions(specItemWinOptions)
+			version := viapi.ParseVersionFromClient(client)
+			windowsOptions, err := flattenWindowsOptions(specItemWinOptions, version)
```
Comment (tenthirtyam) [inline vsphere/internal/helper/guestoscustomizations/customizations_helper.go:335]:
```suggestion
			if version.AtLeast(viapi.VSphereVersion{Product: version.Product, Major: 8, Minor: 2}) {
				windowsOptions, err := flattenWindowsOptions(specItemWinOptions, version)
			}
```

File: vsphere/internal/helper/guestoscustomizations/customizations_helper.go [hunk 3]
```diff
@@ -601,6 +613,7 @@ func expandCustomizationIdentification(d *schema.ResourceData, prefix string) ty
 		JoinWorkgroup: d.Get(prefix + "workgroup").(string),
 		JoinDomain:    d.Get(prefix + "join_domain").(string),
 		DomainAdmin:   d.Get(prefix + "domain_admin_user").(string),
+		DomainOU:      d.Get(prefix + "domain_ou").(string),
```
Comment (tenthirtyam) [inline vsphere/internal/helper/guestoscustomizations/customizations_helper.go:616]:
```suggestion
        	if version.AtLeast(viapi.VSphereVersion{Product: version.Product, Major: 8, Minor: 0, Patch: 2}) {
		DomainOU:      d.Get(prefix + "domain_ou").(string),
	}
```

File: website/docs/r/virtual_machine.html.markdown [hunk 4]
```diff
@@ -1267,6 +1267,25 @@ The options are:
...
+
+```hcl
+resource "vsphere_virtual_machine" "vm" {
+  # ... other configuration ...
+  clone {
+    # ... other configuration ...
+    customize {
+      # ... other configuration ...
+      windows_options {
+        domain_ou             = "OU=Sandbox,OU=Workstations,DC=example,DC=com"
```
Comment (tenthirtyam) [inline website/docs/r/virtual_machine.html.markdown:1282]:
```suggestion
        domain_ou = "OU=Sandbox,OU=Workstations,DC=example,DC=com"
```

File: vsphere/internal/helper/guestoscustomizations/customizations_helper.go (hunk 3 above)
Comment (ultral) [reply vsphere/internal/helper/guestoscustomizations/customizations_helper.go:616]: @tenthirtyam are you sure? there is no `viapi` object in the `expandCustomizationIdentification` function

Comment (ultral) [reply vsphere/internal/helper/guestoscustomizations/customizations_helper.go]: i've implemented that diffidently

```go
	obj := types.CustomizationIdentification{
		JoinWorkgroup: d.Get(prefix + "workgroup").(string),
		JoinDomain:    d.Get(prefix + "join_domain").(string),
		DomainAdmin:   d.Get(prefix + "domain_admin_user").(string),
	}
	if version.AtLeast(viapi.VSphereVersion{Product: version.Product, Major: 8, Minor: 0, Patch: 2}) {
		obj.DomainOU = d.Get(prefix + "domain_ou").(string)
	}
```

File: vsphere/internal/helper/guestoscustomizations/customizations_helper.go [hunk 5]
```diff
@@ -608,12 +610,14 @@ func expandCustomizationGuiUnattended(d *schema.ResourceData, prefix string) typ
...
-func expandCustomizationIdentification(d *schema.ResourceData, prefix string) types.CustomizationIdentification {
+func expandCustomizationIdentification(d *schema.ResourceData, prefix string, version viapi.VSphereVersion) types.CustomizationIdentification {
 	obj := types.CustomizationIdentification{
 		JoinWorkgroup: d.Get(prefix + "workgroup").(string),
 		JoinDomain:    d.Get(prefix + "join_domain").(string),
 		DomainAdmin:   d.Get(prefix + "domain_admin_user").(string),
-		DomainOU:      d.Get(prefix + "domain_ou").(string),
+	}
+	if version.AtLeast(viapi.VSphereVersion{Product: version.Product, Major: 8, Minor: 0, Patch: 2}) {
+		obj.DomainOU = d.Get(prefix + "domain_admin_user").(string)
```
Comment (tenthirtyam) [inline vsphere/internal/helper/guestoscustomizations/customizations_helper.go]: `domain_ou`?

Comment (ultral) [reply vsphere/internal/helper/guestoscustomizations/customizations_helper.go]: Sure, it was in a minute after the wrong commit. Could you please check?

**Comment by ultral** [comment]
> Added comment.

it was in a minute after the wrong commit. Could you please check?

**Comment by ultral** [comment]
@tenthirtyam are there any other required changes?

**Comment by tenthirtyam** [comment]
I'll review in the coming days.

File: website/docs/r/virtual_machine.html.markdown [hunk 6]
```diff
@@ -1267,6 +1267,25 @@ The options are:
...
+
+```hcl
+resource "vsphere_virtual_machine" "vm" {
+  # ... other configuration ...
+  clone {
+    # ... other configuration ...
+    customize {
+      # ... other configuration ...
+      windows_options {
+        domain_ou = "OU=Sandbox,OU=Workstations,DC=example,DC=com"
```
Comment (tenthirtyam) [inline website/docs/r/virtual_machine.html.markdown:1282]:
```suggestion
        domain_ou = "OU=bar,OU=foo,DC=example,DC=com"
```

Comment (tenthirtyam) [reply website/docs/r/virtual_machine.html.markdown]: Resolved in https://github.com/hashicorp/terraform-provider-vsphere/commit/e27590dd9fb5b6dab92e5341c932be6fdd1421ae.

**Review by spacegospod — APPROVED** [review]

**Pull request merged**